src/Compiler/Checking/Expressions/CheckExpressions.fs

```diff
@@ -9686,6 +9686,8 @@ and GetNewInferenceTypeForMethodArg (cenv: cenv) env tpenv x =
     | SynExpr.Lambda (body = a)
     | SynExpr.DotLambda (expr = a) ->
         mkFunTy g (NewInferenceType g) (GetNewInferenceTypeForMethodArg cenv env tpenv a)
+    | SynExpr.MatchLambda _ ->
+        mkFunTy g (NewInferenceType g) (NewInferenceType g)
```

Comment:
Would it make sense to look at the body of the first branch? Or even to continue looking at the other branches if the first branch leads to `NewInferenceType`?

```suggestion
    | SynExpr.Lambda (body = a)
    | SynExpr.DotLambda (expr = a)
    | SynExpr.MatchLambda (matchClauses = SynMatchClause (resultExpr = a) :: _) ->
        mkFunTy g (NewInferenceType g) (GetNewInferenceTypeForMethodArg cenv env tpenv a)
```

I don't know the answer, but it seems like it _could_ affect inference for something like

```fsharp
M (function x -> function y -> function z -> <@ x + y + z @>)
```

or

```fsharp
M (function
    | None -> fun x -> x + 2
    | Some f -> f)
```

Reply:
Let’s add some test to find out 👍🏻